File: modules/rest/src/main/java/dev/httpmarco/polocloud/modules/rest/controller/impl/v1/controller/AuthController.java
package dev.httpmarco.polocloud.modules.rest.controller.impl.v1.controller;

import dev.httpmarco.polocloud.modules.rest.RestModule;
import dev.httpmarco.polocloud.modules.rest.controller.Controller;
import dev.httpmarco.polocloud.modules.rest.controller.impl.v1.model.AuthModel;
import dev.httpmarco.polocloud.modules.rest.controller.methods.Request;
import dev.httpmarco.polocloud.modules.rest.controller.methods.RequestType;
import io.javalin.http.Context;
import io.javalin.http.Cookie;

import java.util.concurrent.TimeUnit;

public class AuthController extends Controller {

    public AuthController(RestModule restModule) {
        super("/auth", restModule);
    }

    @Request(requestType = RequestType.POST, path = "/login")
    public void login(Context context) {
        AuthModel authModel;
        try {
            authModel = context.bodyAsClass(AuthModel.class);
        } catch (Exception e) {
            context.status(400).result(failMessage("Invalid body"));
            return;
        }

        if (authModel.username() == null || authModel.password() == null) {
            context.status(400).result(failMessage("Invalid body: Missing fields"));
            return;
        }

        var token = restModule().userService().login(authModel.username(), authModel.password(), context.ip(), context.userAgent());
        if (token == null) {
            context.status(401).result(failMessage("Invalid credentials"));
            return;
        }

        var cookie = new Cookie("token", token);
        cookie.setHttpOnly(true);
        cookie.setMaxAge((int) TimeUnit.DAYS.toSeconds(7));
        cookie.setSecure(true);

        context.status(200).cookie(cookie);
    }

    // Verify the token
    @Request(requestType = RequestType.GET, path = "/token")
    public void checkToken(Context context) {
        context.status(200); // if the user gets through the auth, the token is valid
    }
}

File: modules/rest/src/main/kotlin/dev/httpmarco/polocloud/modules/rest/controller/impl/v3/controller/AuthController.kt
package dev.httpmarco.polocloud.modules.rest.controller.impl.v3.controller

import com.google.gson.JsonObject
import dev.httpmarco.polocloud.modules.rest.RestModule
import dev.httpmarco.polocloud.modules.rest.auth.user.token.Token
import dev.httpmarco.polocloud.modules.rest.auth.user.User
import dev.httpmarco.polocloud.modules.rest.controller.Controller
import dev.httpmarco.polocloud.modules.rest.controller.defaultResponse
import dev.httpmarco.polocloud.modules.rest.controller.impl.v3.model.AuthModel
import dev.httpmarco.polocloud.modules.rest.controller.methods.Request
import dev.httpmarco.polocloud.modules.rest.controller.methods.RequestType
import io.javalin.http.Context
import io.javalin.http.Cookie
import java.util.concurrent.TimeUnit

class AuthController : Controller("/auth") {

    @Request(requestType = RequestType.POST, path = "/login")
    fun login(context: Context) {
        val model = context.parseBodyOrBadRequest<AuthModel>() ?: return
        if (!context.validate(model.username.isNotBlank(), "Username is required")) return
        if (!context.validate(model.password.isNotBlank(), "Password is required")) return

        val token = RestModule.Companion.instance.userProvider.login(model.username, model.password, context.ip(), context.userAgent())
        if (token == null) {
            context.defaultResponse(401, "Invalid credentials")
            return
        }

        val cookie = Cookie(
            "token",
            token.value,
            maxAge = TimeUnit.DAYS.toSeconds(7).toInt(),
            isHttpOnly = true,
            secure = true,
        )

        context.cookie(cookie).defaultResponse(200, "Login successful")
    }

    @Request(requestType = RequestType.POST, path = "/logout")
    fun logout(context: Context, user: User, token: Token) {
        RestModule.instance.userProvider.logout(user, token)

        context.removeCookie("token")
        context.defaultResponse(200, "Logout successful")
    }

    @Request(requestType = RequestType.GET, path = "/token")
    fun checkToken(context: Context) {
        val token = context.cookie("token")

        if (token == null) {
            context.defaultResponse(401,"No token found")
            return
        }

        val data = JsonObject().apply {
            addProperty("token", token)
        }

        context.defaultResponse(200, "Token is valid", data)
    }
}